source/common/common/token_bucket_impl.cc

```diff
@@ -11,9 +11,16 @@ TokenBucketImpl::TokenBucketImpl(uint64_t max_tokens, TimeSource& time_source, d
 uint64_t TokenBucketImpl::consume(uint64_t tokens, bool allow_partial) {
   if (tokens_ < max_tokens_) {
     const auto time_now = time_source_.monotonicTime();
-    tokens_ = std::min((std::chrono::duration<double>(time_now - last_fill_).count() * fill_rate_) +
-                           tokens_,
-                       max_tokens_);
+
+    const uint64_t new_fill_tokens =
+        std::chrono::duration_cast<std::chrono::microseconds>(time_now - last_fill_).count() *
+            fill_rate_ +
+        last_tokens_;
+    const uint64_t new_tokens = new_fill_tokens / 1000000;
+    const uint64_t last_clac = new_fill_tokens % 1000000;
+    last_tokens_ = last_clac;
+
+    tokens_ = std::min(new_tokens + tokens_, max_tokens_);
```

Comment:
Oops.  I had my mind set on this being an optimization instead of correctness fix.

I can see the double math accumulating error when called many times per second.  The idea behind the suggestions to skip some updates `if (new_fill_tokens < 1000000)` and skip adding tokens `if (tokens < tokens_) {` may help reduce error in computations as long as the consumption rate is relatively steady.

I'll do another pass with error rates in mind.  I think we need some better tests to capture accumulated errors in the original implementation.

Reply:
Yep, optimize is my intention！

_The idea behind the suggestions to skip some updates if (new_fill_tokens < 1000000) and skip adding tokens if (tokens < tokens_) { may help reduce error in computations as long as the consumption rate is relatively steady._

Yes I get it! thanks!

Looking forward to your test！

Do I want to ask whether this pr can continue？ if Ok，I will continue！

Reply:
> I doubt I'll be able to find time in the next week to improve tests for rate limit functionality. I think that you're in a better position to write a test that captures cases where the current implementation accumulates errors and demonstrates how your implementation does better in the scenarios that you have been seeing errors.

okay！I try to do it!

Reply:
> current implementation accumulates



> I doubt I'll be able to find time in the next week to improve tests for rate limit functionality. I think that you're in a better position to write a test that captures cases where the current implementation accumulates errors and demonstrates how your implementation does better in the scenarios that you have been seeing errors.

Yes, I will try to do it. 

 The scenarios that you have been seeing errors that is set 5000 qps, but set fill_rate_ = 3000, the company’s internal display can only limit the current to 2500 instead of close to 3000 by using original token bucket！

So I understand that error is 3000 - 2500 = 500， I want to implement the test in this way, and personally feel that it is not possible to fully reproduce my scene through a single test。

So can I count the error value between the two to confirm that the method is better？